### DISCUSSION TITLE
reading coordinate data

### DISCUSSION THREAD

**Discussion by nfraehr** [DHI/mikeio1d#26] [Q&A] [answered]
Hi

Is there a way to read the xy-coordinates of a node or gridpoint (all locations water level data is calculated)?
For later use in GeoPandas or GIS applications.

**Comment by hawklorry** [comment, marked as answer]
Hi @nfraehr , yes. List below is a function I'm using to get the coordinate of all nodes and create a GeoDataFrame. You will need to get all the node with res1d.data.Nodes. 

`def get_node_geometry(nodes):
    allnodes = [[n.ID, Point(n.XCoordinate, n.YCoordinate)] for n in nodes]
    allnode_df = pd.DataFrame(allnodes, columns=['MUID', 'geometry'])
    allnode_df.set_index('MUID', inplace = True)

    return gpd.GeoDataFrame(allnode_df, geometry = 'geometry')`

**Reply by nfraehr** [reply]
Thanks.

I had found a way, but this is much more neat!

**Reply by Prebbish** [reply]
Hi hawklorry,

I'm still quite new to Python, and I can't get your function to work. 

How should i define the variable Point?

So far I've tried this, but it doesn't work :) 

![Skærmbillede 2022-11-29 113131](https://user-images.githubusercontent.com/119405240/204505976-8b55b4c9-d1a5-40a2-a454-053fa8789250.png)

**Reply by Prebbish** [reply]
I figured out I needed to import the Point function from the shapely package.

"from shapely.geometry import Point"

**Comment by Snowthe** [comment]
Hi,
I was looking for a similar solution myself - however for all points, not only the nodes. Maybe the below code sniplet is helpful:
```
def get_Qpoint_geometry(r1d):
    allps = []
    allreaches = r1d.reaches
    for reach in allreaches:
        subreaches = r1d.reaches[reach]
        if isinstance(subreaches, list):
            for i, subreach in enumerate(subreaches):
                allps.extend([[reach, p.Chainage, p.X, p.Y] for p in r1d.reaches[reach][i].GridPoints.ToArray()])
        else:
            allps.extend([[reach, p.Chainage, p.X, p.Y] for p in r1d.reaches[reach].GridPoints.ToArray()])
    allps_df = pd.DataFrame(allps, columns=['Branch','Chainage','X','Y'])
    # drop duplicates
    allps_df.drop(index=allps_df[allps_df.duplicated(subset=['Branch','Chainage'], keep='last')].index, inplace=True)
    return allps_df
```

...or is there a better/more elegant way of obtaining this information already implemented in mikeio1d?

**Reply by Snowthe** [reply]
Update: With the newer version, since 0.6.0 (or even earlier?), this function has to be adapted:
```
def get_Qpoint_geometry(r1d):
    allps = []
    allreaches = list(r1d.reaches)
    for reach in allreaches:
        subreaches = r1d.reaches[reach]
        if isinstance(subreaches, list):
            for i, subreach in enumerate(subreaches):
                allps.extend([[subreaches[i].name, p.chainage, p.c, p.y] for p in subreaches[0].geometry.points if p.is_gridpoint()])
        else:
            allps.extend([[subreaches.name, p.chainage, p.x, p.y] for p in subreaches.geometry.points if p.is_gridpoint()])
    allps_df = pd.DataFrame(allps, columns=['Branch','Chainage','X','Y'])
    # drop duplicates
    allps_df.drop(index=allps_df[allps_df.duplicated(subset=['Branch','Chainage'], keep='last')].index, inplace=True)
    return allps_df
```
...or does there now exist a more convenient function in-built? I saw some .to_geopandas, but that I could only get me to return the streams as polylines, without the chainage information
